packages/bundlers/experimental/src/ExperimentalBundler.js

```diff
@@ -0,0 +1,949 @@
+            idealBundle.sourceBundles.join(','),
+          needsStableName: idealBundle.needsStableName,
+          bundleBehavior: idealBundle.bundleBehavior,
+          type: idealBundle.type,
+          target: idealBundle.target,
+          env: idealBundle.env,
+        }),
+      );
+    } else {
+      bundle = nullthrows(
+        bundleGraph.createBundle({
+          entryAsset,
+          needsStableName: idealBundle.needsStableName,
+          bundleBehavior: idealBundle.bundleBehavior,
+          target: idealBundle.target,
+        }),
+      );
+    }
+
+    idealBundleToLegacyBundle.set(idealBundle, bundle);
+
+    for (let asset of idealBundle.assets) {
+      bundleGraph.addAssetToBundle(asset, bundle);
+    }
+  }
+  // Step 2: Internalize dependencies for bundles
+  for (let [, idealBundle] of idealBundleGraph.nodes) {
+    if (idealBundle === 'root') continue;
+    let bundle = nullthrows(idealBundleToLegacyBundle.get(idealBundle));
+    for (let internalized of idealBundle.internalizedAssetIds) {
+      let incomingDeps = bundleGraph.getIncomingDependencies(
+        bundleGraph.getAssetById(internalized),
+      );
+      for (let incomingDep of incomingDeps) {
+        if (
+          incomingDep.priority === 'lazy' &&
+          bundle.hasDependency(incomingDep)
+        ) {
+          bundleGraph.internalizeAsyncDependency(bundle, incomingDep);
+        }
+      }
+    }
+  }
+  // Step 3: Add bundles to their bundle groups
+  for (let [bundleId, bundleGroup] of entryBundleToBundleGroup) {
+    let outboundNodeIds = idealBundleGraph.getNodeIdsConnectedFrom(bundleId);
+    for (let id of outboundNodeIds) {
+      let siblingBundle = nullthrows(idealBundleGraph.getNode(id));
+      invariant(siblingBundle !== 'root');
+      let legacySiblingBundle = nullthrows(
+        idealBundleToLegacyBundle.get(siblingBundle),
+      );
+      bundleGraph.addBundleToBundleGroup(legacySiblingBundle, bundleGroup);
+    }
+  }
+
+  // Step 4: Add references to all bundles
+  for (let [asset, references] of idealGraph.assetReference) {
+    for (let [dependency, bundle] of references) {
+      let legacyBundle = nullthrows(idealBundleToLegacyBundle.get(bundle));
+      bundleGraph.createAssetReference(dependency, asset, legacyBundle);
+    }
+  }
+}
+
+function createIdealGraph(
+  assetGraph: MutableBundleGraph,
+  config: ResolvedBundlerConfig,
+): IdealGraph {
+  // Asset to the bundle and group it's an entry of
+  let bundleRoots: Map<BundleRoot, [NodeId, NodeId]> = new Map();
+  let bundles: Map<string, NodeId> = new Map();
+  let dependencyBundleGraph: DependencyBundleGraph = new ContentGraph();
+  let assetReference: DefaultMap<
+    Asset,
+    Array<[Dependency, Bundle]>,
+  > = new DefaultMap(() => []);
+
+  // bundleRoot to all bundleRoot descendants
+  let reachableBundles: DefaultMap<
+    BundleRoot,
+    Set<BundleRoot>,
+  > = new DefaultMap(() => new Set());
+
+  let bundleGraph: Graph<Bundle | 'root'> = new Graph();
```

Comment:
(ideal) bundleGraph has a root node that is connected to all entry (entries of the application) bundles so that later, when we remove nodes and remove edges, **no clean up** occurs. 

`removeEdge` contains clean up for orphans, which we **never** want. This graph is maximally deduplicated, and an "island"  only signifies that a bundle is the only bundle in it's "bundle Group", since there are no bundle groups. So there should never be an orphans that we don't already handle